File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/NoExceptionsInFinally.cs
using System;
using System.Collections.Generic;

namespace Tests.Diagnostics
{
    public class MyTestCases
    {
        public void Method1()
        {
            try
            {
                throw new ArgumentException();
            }
            finally
            {
                throw new InvalidOperationException(); // Noncompliant
            }
        }

        public void Method2()
        {
            try
            {
                throw new ArgumentException();
            }
            finally
            {
                try
                {

                }
                finally
                {
                    throw new InvalidOperationException(); // Noncompliant
                }
            }
        }

        public void Method3()
        {
            try
            {
                throw new ArgumentException();
            }
            finally
            {
                try
                {
                    throw new InvalidOperationException(); // Noncompliant
                }
                catch (InvalidCastException)
                {

                }
            }
        }

        public void Method4()
        {
            int a = 0;
            try
            {
                a++;
                throw new ArgumentException();
            }
            finally
            {
                if (a > 0)
                {
                    throw new InvalidOperationException(); // Noncompliant
                }
            }
        }

        public void Method5()
        {
            try
            {
                throw new ArgumentException();
            }
            finally
            {
                try
                {
                    try
                    {
                        throw new ArgumentOutOfRangeException(); // Noncompliant
                    }
                    finally
                    {
                        throw new InvalidOperationException(); // Noncompliant
                    }
                }
                catch
                {

                }
            }
        }

        public void Method6()
        {
            try
            {
                throw new ArgumentException();
            }
            finally
            {
                try
                {
                    try
                    {
                        throw new InvalidOperationException(); // Noncompliant
                    }
                    catch
                    {

                    }

                    throw new ArgumentOutOfRangeException(); // Noncompliant
                }
                catch (ArgumentOutOfRangeException)
                {
                    throw; // Noncompliant
                }
            }
        }

        public void Method7()
        {
            try
            {
                throw new ArgumentException();
            }
            finally
            {
                try
                {
                    try
                    {
                        try
                        {
                            throw new InvalidOperationException(); // Noncompliant
                        }
                        catch
                        {

                        }

                        throw new ArgumentOutOfRangeException(); // Noncompliant
                    }
                    catch (ArgumentOutOfRangeException)
                    {
                        throw; // Noncompliant
                    }
                }
                catch (ArgumentOutOfRangeException)
                {
                }
            }
        }

        public void Method8()
        {
            try
            {
                throw new ArgumentException();
            }
            finally
            {
            }
        }

        public void Method9()
        {
            try
            {
                throw new ArgumentException();
            }
            finally
            {
                try
                {
                    throw new InvalidOperationException(); // Noncompliant
                }
                catch (InvalidOperationException)
                {

                }
            }
        }

        public void Method10()
        {
            try
            {
                throw new ArgumentException();
            }
            finally
            {
                try
                {
                    throw new InvalidOperationException(); // Noncompliant
                }
                catch
                {

                }
            }
        }

        public void Method11()
        {
            try
            {
                throw new ArgumentException();
            }
            finally
            {
                try
                {
                    try
                    {
                        try
                        {
                            throw new InvalidOperationException(); // Noncompliant
                        }
                        catch
                        {

                        }

                        throw new ArgumentOutOfRangeException(); // Noncompliant
                    }
                    catch (ArgumentOutOfRangeException)
                    {
                        throw; // Noncompliant
                    }
                }
                catch
                {
                }
            }
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/NoExceptionsInFinally.vb
Imports System

Public Class TestCases

    Public Sub Method1()
        Try
            Throw New ArgumentException()
        Finally
            Throw New InvalidOperationException() ' Noncompliant
        End Try
    End Sub

    Public Sub Method2()
        Try
            Throw New ArgumentException()
        Finally
            Try

            Finally
                Throw New InvalidOperationException() ' Noncompliant
            End Try
        End Try
    End Sub

    Public Sub Method3()
        Try
            Throw New ArgumentException()
        Finally
            Try
                Throw New InvalidOperationException() ' Noncompliant
            Catch ex As InvalidCastException
            End Try
        End Try
    End Sub

    Public Sub Method4()
        Dim a As Integer = 0
        Try
            a += 1
            Throw New ArgumentException()
        Finally
            If a > 0 Then Throw New InvalidOperationException() ' Noncompliant
        End Try
    End Sub

    Public Sub Method5()
        Try
            Throw New ArgumentException()
        Finally
            Try
                Try
                    Throw New ArgumentOutOfRangeException() ' Noncompliant
                Finally
                    Throw New InvalidOperationException()   ' Noncompliant
                End Try
            Catch
            End Try
        End Try
    End Sub

    Public Sub Method6()
        Try
            Throw New ArgumentException()
        Finally
            Try
                Try
                    Throw New InvalidOperationException() ' Noncompliant
                Catch
                End Try
                Throw New ArgumentOutOfRangeException() ' Noncompliant
            Catch ex As ArgumentOutOfRangeException
                Throw ' Noncompliant
            End Try
        End Try
    End Sub

    Public Sub Method7()
        Try
            Throw New ArgumentException()
        Finally
            Try
                Try
                    Try
                        Throw New InvalidOperationException() ' Noncompliant
                    Catch
                    End Try
                    Throw New ArgumentOutOfRangeException() ' Noncompliant
                Catch ex As ArgumentOutOfRangeException
                    Throw ' Noncompliant
                End Try
            Catch ex As ArgumentOutOfRangeException
            End Try
        End Try
    End Sub

    Public Sub Method8()
        Try
            Throw New ArgumentException()
        Finally
        End Try
    End Sub

    Public Sub Method9()
        Try
            Throw New ArgumentException()
        Finally
            Try
                Throw New InvalidOperationException() ' Noncompliant
            Catch ex As InvalidOperationException
            End Try
        End Try
    End Sub

    Public Sub Method10()
        Try
            Throw New ArgumentException()
        Finally
            Try
                Throw New InvalidOperationException() ' Noncompliant
            Catch
            End Try
        End Try
    End Sub

    Public Sub Method11()
        Try
            Throw New ArgumentException()
        Finally
            Try
                Try
                    Try
                        Throw New InvalidOperationException() ' Noncompliant
                    Catch
                        Throw New ArgumentOutOfRangeException() ' Noncompliant
                    End Try
                Catch ex As ArgumentOutOfRangeException
                    Throw ' Noncompliant
                End Try
            Catch
            End Try
        End Try
    End Sub

End Class
